### PR TITLE
Mautic

### PR DESCRIPTION
**Pull request by jospoortvliet**
WIP integrating Mautic.

TODO:

- [x] figure out how to submit stuff in Mautic (there go two weeks of my life)
- [x] use for all pages:
  - [x] healthcare vertical
  - [x] security
  - [x] education vertical
  - [x] enterprise page
  - [x] storage
  - [x] user management
  - [x] monitoring
  - [x] endtoend
  - [x] file-access-control
  - [x] globalscale
  - [x] nextcloud-12 wp
Created new whitepaper page as well with all of them.

This only works when a cron job kicks off the mautic-authenticate.php script every ~hour. Also, for security we have to make sure the file it creates (mautictoken.json) can't be retrieved from the outside!

Billions of thanks to @camilasan and @Espina2 as of course I could never do this without their help... Many many hours from both... 

Time to test this.

- [ ] Working locally:
  - [ ] healthcare vertical
  - [ ] security
  - [ ] education vertical
  - [ ] enterprise page
  - [ ] storage
  - [ ] user management
  - [ ] monitoring
  - [ ] endtoend
  - [ ] file-access-control
  - [ ] globalscale
  - [ ] nextcloud-12 wp
  - [ ] new whitepaper page
     - [x] architecture
     - [x] security
     - [x] server-side enc
     - [x] tu berlin
     - [x] end-to-end
     - [x] NCC group review
     - [x] global scale
     - [x] health care
     - [x] GPDPR
     - [x] monitoring
     - [x] global workflow
     - [x] ldap
     - [x] wnd
     - [x] sharepoint

As we have no test infra, we have to test this in live (...) and then decide to roll back. Fun.
- [x] Working live:
  - [ ] healthcare vertical
  - [ ] security
  - [ ] education vertical
  - [ ] enterprise page
  - [ ] storage
  - [ ] user management
  - [ ] monitoring
  - [ ] endtoend
  - [ ] file-access-control
  - [ ] globalscale
  - [ ] nextcloud-12 wp
  - [x] new whitepaper page
     - [x] architecture
     - [ ] security
     - [ ] server-side enc
     - [ ] tu berlin
     - [ ] end-to-end
     - [ ] NCC group review
     - [ ] global scale
     - [ ] health care
     - [ ] GPDPR
     - [ ] monitoring
     - [ ] global workflow
     - [ ] ldap
     - [ ] wnd
     - [ ] sharepoint

### CODE REVIEW

File: mautic-authenticate.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            var_dump($accessTokenData);
+
+            $file = './mautictoken.json';
+            $content = json_encode($accessTokenData);
+            file_put_contents($file, $content);
+        }
+    }
+} catch (Exception $e) {
+    // Do Error handling
+var_dump($e);
```
Comment (MorrisJobke) [inline mautic-authenticate.php]: Nope 🙈

File: mautic-authenticate.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        // Obtain the access token returned; call accessTokenUpdated() to catch if the token was updated via a
+        // refresh token
+
+        // $accessTokenData will have the following keys:
+        // For OAuth1.0a: access_token, access_token_secret, expires
+        // For OAuth2: access_token, expires, token_type, refresh_token
+
+        if ($auth->accessTokenUpdated()) {
+            $accessTokenData = $auth->getAccessTokenData();
+            var_dump($accessTokenData);
```
Comment (MorrisJobke) [inline mautic-authenticate.php]: 🙈

File: mautic-submit.php [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+        // remove the contact from the segment if they're in it. This allows them to get the email again. -> this doesn't work in tests, asked Mautic support if they have any other suggestions...
+        $response = $segmentApi->removeContact($segmentId, $contactId);
+        // ignore errors in the above... don't care if the person wasn't in there
+        $response = '';
+        // add the contact to the segment (this will typically result in an action being taken, like a mail sent)
+        $response = $segmentApi->addContact($segmentId, $contactId);
+        if (!isset($response['success'])) {
+            $error_message .= 'Adding user to segment failed, error:<br />';
+            $error_message .= $response . '.<br />';
```
Comment (MorrisJobke) [inline mautic-submit.php]: XSS vector if their API responses with weird stuff. Should not be shown to the end user. Either log this or remove it.

**Comment by MorrisJobke** [comment]
Improved the mautic submit code a lot ;) https://github.com/nextcloud/nextcloud.com/pull/802/files#diff-2f5fab86a2b5b6d8b6f4976596d374a8

It now uses basic auth (there is no other way to authenticate against mautic API as of now without Oauth, which is clearly not what we want, because redirecting website users to mautic to log in there is not useful 😉 )

**Comment by MorrisJobke** [comment]
@jospoortvliet The resend of email does not work when I tested it. Is there a timeout for the resend configured?

**Comment by jospoortvliet** [comment]
@MorrisJobke I couldn't get it to work either, I've asked support if this should work and if so, what/how/etc.

The manual way is to remove the contact completely from mautic and re-add but that won't work in practice as you lose all other data, too.

**Comment by MorrisJobke** [comment]
> @MorrisJobke I couldn't get it to work either, I've asked support if this should work and if so, what/how/etc.

Ah okay ... then let's drop it from the code as well to not have non-working code there.

**Pull request merged**